src/main/java/net/openhft/chronicle/queue/impl/single/SingleChronicleQueueExcerpts.java

```diff
@@ -862,6 +858,87 @@ public String toString() {
+            @Override
+            public boolean onOpen(boolean metaData, int safeLength) {
+                for (int i = 0; i < 128; i++) {
+                    try {
+                        assert wire != null;
+                        long pos = store.writeHeader(wire, Wires.UNKNOWN_LENGTH, safeLength, timeoutMS());
+                        position(pos);
+                        context.isClosed = false;
+                        context.wire = wire; // Jvm.isDebug() ? acquireBufferWire() : wire;
+                        context.padToCacheAlign = padToCacheAlignMode() != Padding.NEVER;
+                        context.metaData(metaData);
+                        return true;
+
+                    } catch (EOFException theySeeMeRolling) {
+                        cycle = handleRoll(cycle);
+                    }
+                }
+                throw new IllegalStateException("Unable to roll to the current cycle");
+            }
+
+            @Override
+            public void onClose() {
+                // do nothing
+            }
+        }
+
+        /**
+         * Buffered writes while awaiting for other appenders to release the header #403
+         */
+        private class HeaderWriteStrategyDefer implements HeaderWriteStrategy {
+            @Override
+            public boolean onOpen(boolean metaData, int safeLength) {
+                assert wire != null;
+                long pos = store.tryWriteHeader(wire, Wires.UNKNOWN_LENGTH, safeLength);
+                if (pos != WireOut.TRY_WRITE_HEADER_FAILED) {
+                    position(pos);
+                    context.wire = wire; // Jvm.isDebug() ? acquireBufferWire() : wire;
+                    context.deferredHeader = false;
+                } else {
+                    context.wire = acquireBufferWire();
+                    context.deferredHeader = true;
+                }
+
+                context.isClosed = false;
+                context.padToCacheAlign = padToCacheAlignMode() != Padding.NEVER;
+                context.metaData(metaData);
+                return true;
+            }
+
+            public void onClose() {
+                if (context.deferredHeader) {
+                    int safeLength = (int) queue.overlapSize();
+                    assert wire != null;
+                    assert wire != context.wire;
+                    for (int i = 0; i < 128; i++) {
+                        try {
+                            // TODO: we should be able to write and update the header in one go
+                            long pos = store.writeHeader(wire, Wires.UNKNOWN_LENGTH, safeLength, timeoutMS());
+                            position(pos);
+                            long length = context.wire.bytes().copyTo(wire.bytes());
+                            wire.bytes().writePosition(length + wire.bytes().writePosition());
+                            context.wire = wire;
```

Comment:
Should you reset the context wire in a finally block? In the event of an exception, you'll be left using the instance returned from acquireBufferWire()?

Reply:
This method is called when we close the DocumentContext, typically in at the end of a try-with-resources, so it _shouldn't_ be used any more. When we open the DocumentContext, we set context.wire. Does that address your concerns?

Reply:
No, that's fine, if it's the terminal state for that instance, then there's no need for any change.